Check for existing rebuilds

chrismytton opened the pull request:
This checks for existing jobs before re-queueing them, which should help keep the size of the rebuild queue down.

Fixes #37 

# Note to merger

Make sure the following have been done before merging this:

- [x] Change the base of this pull request to `master` once #53 has been merged.

chrismytton committed:
Skip rebuilds for legislatures that are already queued

This checks for an existing job matching the rebuild arguments. If one
is found then the rebuild is skipped to stop the queue from bloating.

tmtmtmtm reviewed (changes requested):
There's really two distinct types things going on here, so I think it would be worth splitting the PR in two — the first for the refactoring (which looks like it could go through like this OK) and then the second for changing the behaviour (which I think also really needs a test to show the new behaviour is working).

chrismytton commented:
> There's really two distinct types things going on here, so I think it would be worth splitting the PR in two — the first for the refactoring (which looks like it could go through like this OK) and then the second for changing the behaviour (which I think also really needs a test to show the new behaviour is working).

Yep, good shout. I've pulled out the refactoring into a separate PR - https://github.com/everypolitician/rebuilder/pull/55.

chrismytton committed:
Re-write tests to use Sidekiq's API

The Sidekiq API doesn't work with Sidekiq's testing mode and we want to
use the API to check for existing jobs, so switching the tests over to
use the API seems like the simplest option.

chrismytton committed:
Add a Rebuilder::Queue class for adding rebuild jobs

This class is responsible for ensuring that jobs for the same
legislature don't get queued up more than once.

tmtmtmtm reviewed (commented):
I'm a little concerned that if I remove the `if` clause in the new `add` method — i.e. making it simply

```ruby
    def add(country, legislature, source = nil)
      RebuilderJob.perform_async(country, legislature, source)
      "Queued rebuild for country=#{country} legislature=#{legislature} source=#{source}\n"
    end
```

then the tests still pass…

This definitely seems to be the right abstraction, but I think finding it at the _end_ of this PR makes it a little harder to follow. I'm wondering if it might make more sense to do this as two steps — firstly a 'clean' refactoring that maintains the old behaviour, but with the new Queue implementation. Then once we know that's working correctly, adding the 'skip if one exists' logic.
